### PR TITLE
Implement conditional remediation

### LINKED ISSUE #41: Conditional remediation on Helm actions
The idea is to add `<Action>Remediation` structs to the API, to be implemented by the Helm actions that should offer a remediation strategy in case they fail (`Install` and `Upgrade`).

### Install

```diff
diff --git a/docs/spec/v2alpha1/helmreleases.md b/docs/spec/v2alpha1/helmreleases.md
index bf6010d..127c196 100644
--- a/docs/spec/v2alpha1/helmreleases.md
+++ b/docs/spec/v2alpha1/helmreleases.md
@@ -96,6 +96,21 @@ type HelmChartTemplate struct {
 	Interval *metav1.Duration `json:"interval,omitempty"`
 }
 
+type InstallRemediation struct {
+	// Retries is the number of retries that should be attempted on failures before
+	// bailing. Defaults to '0', a negative integer equals to unlimited retries.
+	// +optional
+	Retries int `json:"retries,omitempty"`
+
+	// IgnoreTestFailures tells the controller to skip remediation when
+	// the Helm tests are run after an install action but fail.
+	// Defaults to 'Test.IgnoreTestFailures'.
+	// +optional
+	IgnoreTestFailures bool `json:"ignoreTestFailures,omitempty"`
+
+	// SkipLastRemediation tells the controller to skip remediation when
+	// there are no retries left so the release can be debugged.
+	// +optional
+	SkipLastRemediation bool `json:"skipLastRemediation,omitempty"`
+}
```

```diff
diff --git a/docs/spec/v2alpha1/helmreleases.md b/docs/spec/v2alpha1/helmreleases.md
index bf6010d..127c196 100644
 // Install holds the configuration for Helm install actions.
 type Install struct {
 	// Timeout is the time to wait for any individual Kubernetes operation (like Jobs
@@ -104,6 +119,12 @@ type Install struct {
 	// +optional
 	Timeout *metav1.Duration `json:"timeout,omitempty"`

+	// Remediation holds the remediation configuration for when the
+	// Helm install action fails. The default install failure
+	// remediation is an uninstall action.
+	// +optional
+	Remediation InstallRemediation `json:"remediation,omitempty"`
+
 	// DisableWait disables the waiting for resources to be ready after a
 	// Helm install has been performed.
 	// +optional
```

### Upgrade

```diff
diff --git a/docs/spec/v2alpha1/helmreleases.md b/docs/spec/v2alpha1/helmreleases.md
index bf6010d..127c196 100644
--- a/docs/spec/v2alpha1/helmreleases.md
+++ b/docs/spec/v2alpha1/helmreleases.md
@@ -96,6 +96,21 @@ type HelmChartTemplate struct {
 	Interval *metav1.Duration `json:"interval,omitempty"`
 }
 
+type UpgradeRemediation struct {
+	// Retries is the number of retries that should be attempted on failures before
+	// bailing. Defaults to '0', a negative integer equals to unlimited retries.
+	// +optional
+	Retries int `json:"retries,omitempty"`
+
+	// Strategy to use for the remediation of a failed action.
+	// +kubebuilder:validation:Enum=rollback,uninstall
+	// +optional
+	Strategy string `json:"strategy,omitempty"`
+
+	// IgnoreTestFailures tells the controller to skip remediation when
+	// the Helm tests are run after an upgrade action but fail.
+	// Defaults to 'Test.IgnoreTestFailures'.
+	// +optional
+	IgnoreTestFailures bool `json:"ignoreTestFailures,omitempty"`
+
+	// SkipLastRemediation tells the controller to skip remediation when
+	// there are no retries left so the release can be debugged.
+	// +optional
+	SkipLastRemediation bool `json:"skipLastRemediation,omitempty"`
+}
```

```diff
diff --git a/docs/spec/v2alpha1/helmreleases.md b/docs/spec/v2alpha1/helmreleases.md
index bf6010d..127c196 100644
@@ -137,10 +158,11 @@ type Upgrade struct {
 	// +optional
 	Timeout *metav1.Duration `json:"timeout,omitempty"`

-	// MaxRetries is the number of retries that should be attempted on failures before
-	// bailing. Defaults to '0', a negative integer equals to unlimited retries.
+	// Remediation holds the remediation configuration for when the
+	// Helm upgrade action for the HelmRelease fails. The default
+	// upgrade failure remediation is to not perform any action.
 	// +optional
-	MaxRetries int `json:"maxRetries,omitempty"`
+	Remediation UpgradeRemediation `json:"remediation,omitempty"`

 	// DisableWait disables the waiting for resources to be ready after a
 	// Helm upgrade has been performed.
```

### Test

```diff
diff --git a/docs/spec/v2alpha1/helmreleases.md b/docs/spec/v2alpha1/helmreleases.md
index bf6010d..127c196 100644
@@ -179,6 +201,12 @@ type Test struct {
 	// +optional
 	Enable bool `json:"enable,omitempty"`

+	// IgnoreTestFailures tells the controller to skip remediation when
+	// the Helm tests are run but fail.
+	// Can be overwritten for tests run after install or upgrade actions
+	// in 'Install.IgnoreTestFailures' and 'Upgrade.IgnoreTestFailures'.
+	// +optional
+	IgnoreTestFailures bool `json:"ignoreTestFailures,omitempty"`

 	// Timeout is the time to wait for any individual Kubernetes operation
 	// during the performance of a Helm test action. Defaults to
 	// 'HelmReleaseSpec.Timeout'.
```

### Rollback

```diff
diff --git a/docs/spec/v2alpha1/helmreleases.md b/docs/spec/v2alpha1/helmreleases.md
index bf6010d..127c196 100644
@@ -188,11 +216,6 @@ type Test struct {

 // Rollback holds the configuration for Helm rollback actions.
 type Rollback struct {
-	// Enable enables Helm rollback actions for this release after an
-	// Helm install or upgrade action failure.
-	// +optional
-	Enable bool `json:"enable,omitempty"`
-
 	// Timeout is the time to wait for any individual Kubernetes operation (like Jobs
 	// for hooks) during the performance of a Helm rollback action. Defaults to
 	// 'HelmReleaseSpec.Timeout'.
```

## Behavioral notes

* Currently test failures are ignored in the POC and result in `Ready==True`, this behavior changes due to the `IgnoreTestFailures` flags.
* With the `Upgrade.Remediation.Strategy` value of `uninstall`, the retry will be an install. This means that after a successful uninstall, the install remediation configuration takes over.

## Discussions / references

* https://github.com/fluxcd/toolkit/discussions/102
* https://github.com/fluxcd/helm-controller/issues/5

### PR DESCRIPTION
**Pull request by seaneagan**
Fixes #41

### CODE REVIEW

File: api/v2alpha1/helmrelease_types.go [hunk 1]
```diff
@@ -192,10 +275,11 @@ type Upgrade struct {
...
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 
-	// MaxRetries is the number of retries that should be attempted on failures before
-	// bailing. Defaults to '0', a negative integer equals to unlimited retries.
+	// Remediation holds the remediation configuration for when the
+	// Helm upgrade action for the HelmRelease fails. The default
+	// is to not perform any action.
 	// +optional
-	MaxRetries int `json:"maxRetries,omitempty"`
+	Remediation UpgradeRemediation `json:"remediation,omitempty"`
```
[the remark this answers is not in the document]
Comment (seaneagan) [reply api/v2alpha1/helmrelease_types.go]: Done.

File: api/v2alpha1/helmrelease_types.go [hunk 2]
```diff
@@ -148,6 +160,12 @@ type Install struct {
 	// +optional
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 
+	// Remediation holds the remediation configuration for when the
+	// Helm install action for the HelmRelease fails. The default
+	// is to not perform any action.
+	// +optional
+	Remediation InstallRemediation `json:"remediation,omitempty"`
```
[the remark this answers is not in the document]
Comment (seaneagan) [reply api/v2alpha1/helmrelease_types.go]: Done.

File: api/v2alpha1/helmrelease_types.go [hunk 3]
```diff
@@ -412,88 +584,43 @@ func SetHelmReleaseCondition(hr *HelmRelease, condition string, status corev1.Co
...
-	if hr.Spec.Test.Enable {
-		for _, c := range hr.Status.Conditions {
-			if c.Status == corev1.ConditionTrue && (c.Type == InstalledCondition || c.Type == UpgradedCondition) {
-				return true
-			}
-		}
+// HelmReleaseAttempted registers an attempt of the given HelmRelease with the given state,
+// resetting any status related to attempts on old states. It returns the modified HelmRelease
+// and a boolean indicating a state change.
+func HelmReleaseAttempted(hr HelmRelease, revision string, releaseRevision int, valuesChecksum string) (HelmRelease, bool) {
```
[the remark this answers is not in the document]
Comment (seaneagan) [reply api/v2alpha1/helmrelease_types.go:607]: Better now?

[the remark this answers is not in the document]
Comment (hiddeco) [reply api/v2alpha1/helmrelease_types.go:607]: Perfect. It needs a rebase, and when you changed the methods of the interface the actual implementations were not renamed, causing the build to fail.

**Comment by hiddeco** [comment]
Tidied it up a bit more @seaneagan, and fixed some panics due to the pointer changes. Some of the end to end tests are still failing, otherwise it looks good to go to me.

**Review by hiddeco — APPROVED** [review]
Can not thank you enough for your work on this @seaneagan :1st_place_medal: :tulip:

**Pull request merged**